GVFS/GVFS.Common/WindowsProductUpgraderPlatformStrategy.cs

```diff
@@ -0,0 +1,75 @@
+using GVFS.Common.FileSystem;
+using GVFS.Common.Tracing;
+using System;
+using System.IO;
+
+namespace GVFS.Common
+{
+    public class WindowsProductUpgraderPlatformStrategy : ProductUpgraderPlatformStrategy
+    {
+        public WindowsProductUpgraderPlatformStrategy(PhysicalFileSystem fileSystem, ITracer tracer)
+        : base(fileSystem, tracer)
+        {
+        }
+
+        public override bool TryPrepareLogDirectory(out string error)
+        {
+            // Under normal circumstances
+            // ProductUpgraderInfo.GetLogDirectoryPath will have
+            // already been created by GVFS.Service.  If for some
+            // reason it does not (e.g. the service failed to start),
+            // we need to create
+            // ProductUpgraderInfo.GetLogDirectoryPath() explicity to
+            // ensure that it has the correct ACLs (so that both admin
+            // and non-admin users can create log files).  If the logs
+            // directory does not already exist, this call could fail
+            // when running as a non-elevated user.
+            string createDirectoryError;
+            if (!this.FileSystem.TryCreateDirectoryWithAdminAndUserModifyPermissions(ProductUpgraderInfo.GetLogDirectoryPath(), out createDirectoryError))
+            {
+                error = $"ERROR: Unable to create directory `{ProductUpgraderInfo.GetLogDirectoryPath()}`";
+                error += $"\n{createDirectoryError}";
+                error += $"\n\nTry running {GVFSConstants.UpgradeVerbMessages.GVFSUpgrade} from an elevated command prompt.";
+                return false;
+            }
+
+            error = null;
+            return true;
+        }
+
+        public override bool TryPrepareApplicationDirectory(out string error)
+        {
+            string rootDirectoryPath = ProductUpgraderInfo.GetUpgradesDirectoryPath();
+            string toolsDirectoryPath = Path.Combine(rootDirectoryPath, ProductUpgrader.ToolsDirectory);
+
+            Exception deleteDirectoryException;
+            if (this.FileSystem.DirectoryExists(toolsDirectoryPath) &&
+                !this.FileSystem.TryDeleteDirectory(toolsDirectoryPath, out deleteDirectoryException))
+            {
+                error = $"Failed to delete {toolsDirectoryPath} - {deleteDirectoryException.Message}";
+
+                /// this.TraceException(deleteDirectoryException, nameof(this.TrySetupToolsDirectory), $"Error deleting {toolsDirectoryPath}.");

```

Comment:
Is this intended to be commented-out?

Reply:
I was lazy for my initial round of changes. The base class did not have the `TraceException` functionality. I added this to the base class, and uncommented this (and similar) lines.